### PR TITLE
Update google-api-client and Google agents

### PR DESCRIPTION
**Pull request by albertsun**
This PR started with trying to implement a GoogleBigqueryAgent as a separate gem, using https://rubygems.org/gems/google-cloud-bigquery/

Sadly, that led to a bunch of gem dependency conflicts, which this PR tries to untangle by updating various gems. Most of them were no problem, except for `google-api-client`. https://github.com/google/google-api-ruby-client/blob/master/MIGRATING.md

Two existing agents use that. GoogleTranslateAgent was able to be easily upgraded, it nicely follows the new Cloud API pattern (so there's a new gem for that, but it's all backwards compatible).

GoogleCalendarPublishAgent was a bit more painful to keep it working with service accounts.

This breaks backwards compatibility for the GoogleCalendarPublishAgent for people using PKCS12 keys, due to Google's new preferred authentication method. PKCS12 keys are no longer supported by the latest googleauth gem, they say to use JSON key files / RSA private keys.

The new API also no longer uses the API discovery hoopla. It implements all API calls as Ruby methods, and converts keys to snake case. So `dateTime` becomes `date_time`. `time_zone` also appears to now be required.

I'm not sure how many people are using that agent, or if others have ideas for how to smooth the upgrade for them. I think the new APIs are nicer, but doesn't appear to be any automated way to upgrade people's authentication. Maybe @CloCkWeRX you have some ideas from writing the initial agent? Or anyone else?

### CODE REVIEW

File: lib/google_calendar.rb [hunk 1]
```diff
@@ -1,33 +1,37 @@
...
+
 class GoogleCalendar
   def initialize(config, logger)
     @config = config
 
     if @config['google']['key'].present?
-      @key = OpenSSL::PKCS12.new(@config['google']['key'], @config['google']['key_secret']).key
+      # https://github.com/google/google-auth-library-ruby/issues/65
+      # https://github.com/google/google-api-ruby-client/issues/370
+      ENV['GOOGLE_PRIVATE_KEY'] = @config['google']['key']
```
Comment (dsander) [inline lib/google_calendar.rb]: Is it possible to authenticate without using the environment? As far as I know the `ENV` persists for the current process and 'leaks' to all other Agents.

File: Procfile [hunk 2]
```diff
@@ -5,6 +5,7 @@
 # Procfile for development using the new threaded worker (scheduler, twitter stream and delayed job)
 web: bundle exec rails server -p ${PORT-3000} -b ${IP-0.0.0.0}
 jobs: bundle exec rails runner bin/threaded.rb
+dj: bundle exec script/delayed_job run
```
Comment (dsander) [inline Procfile]: Is this change required? `DelayedJob` should already be started in the threaded background process.

File: Gemfile [hunk 3]
```diff
@@ -44,19 +44,23 @@ gem 'slack-notifier', '~> 1.0.0'  # SlackAgent
...
 # Weibo Agents
-# FIXME needs to loosen omniauth dependency
-gem 'weibo_2', github: 'dsander/weibo_2', branch: 'master'
+# FIXME needs to loosen omniauth dependency, add rest-client
+gem 'weibo_2', github: 'albertsun/weibo_2', branch: 'master'
 
 # GoogleCalendarPublishAgent and GoogleTranslateAgent
-gem 'google-api-client', '~> 0.7.1', require: 'google/api_client'
+gem 'google-api-client', '~> 0.13'
+# gem 'google-api-client', '~> 0.7.0' #, require: 'google/api_client'
```
Comment (dsander) [inline Gemfile]: Can we remove this?

File: Gemfile [hunk 4]
```diff
@@ -44,19 +44,23 @@ gem 'slack-notifier', '~> 1.0.0'  # SlackAgent
...
+gem 'google-cloud-translate', '~> 1.0.0', require: 'google/cloud/translate'
+# gem 'google-cloud-bigquery', '~> 0.27.0', require: 'google/cloud/bigquery'
```
Comment (dsander) [inline Gemfile]: Same here, I guess you would add this in your gem, right?

File: Procfile (hunk 2 above)
Comment (albertsun) [inline Procfile]: Not needed. I was having some trouble with things running in a timely manner in my local env so added this and forgot to take out.

File: Gemfile (hunk 4 above)
Comment (albertsun) [inline Gemfile]: Yea, can take these out. Unless people would be interested in having the bigquery agent in core.

File: lib/google_calendar.rb (hunk 1 above)
Comment (albertsun) [inline lib/google_calendar.rb]: I wasn't able to figure out how to do that, while continuing to authenticate with a service account. Good point about it leaking. Would adding a cleanup method to remove those environment variables before the end of `receive` be safe enough?

**Comment by albertsun** [comment]
Upgrade Warning is a nice feature. Will add a section there.

For converting the event formats. Are you recommending adding some compatibility code in the `receive` method? Or some sort of migration that runs against all agents upstream of a GoogleCalendarPublishAgent to fix the keys they are using? The latter seems like it might be better, but it'll be difficult to do in a robust way because there could be so many varied sources / other agents that depend on the former option format (i.e. an upstream agent that goes to both a google cal one, and some unrelated one that's expecting `dateTime` now)

File: lib/google_calendar.rb (hunk 1 above)
Comment (dsander) [inline lib/google_calendar.rb]: That's a bummer, unsetting them would certainly help. The only other option would be to change the authentication to OAuth2, right? But that would make the migration even more annoying.

**Comment by dsander** [comment]
> Are you recommending adding some compatibility code in the receive method?

Yes, I was thinking about that but then checked [all possible arguements](https://developers.google.com/google-apps/calendar/v3/reference/events/insert) so maybe that would be overkill.

**Comment by albertsun** [comment]
Yea, I think that could lead to things being even more unpredictable. Could we come up with a better way of communicating backwards compatibility changes?

I added one check for the most common use case here, of `dateTime`. Also turns out `time_zone` is not actually necessary, so removed that from the default options.

**Comment by albertsun** [comment]
Cool, thanks @dsander !

Just pushed another commit to make the upgrade warning a little clearer.

Also, if you're curious, here is the agent gem for Google BigQuery: https://github.com/albertsun/huginn_bigquery_agent

File: app/models/agents/google_calendar_publish_agent.rb [hunk 5]
```diff
@@ -95,14 +109,27 @@ def receive(incoming_events)
...
+            )
 
         create_event :payload => {
           'success' => true,
           'published_calendar_event' => calendar_event,
           'agent_id' => event.agent_id,
           'event_id' => event.id
         }
+
+        calendar.cleanup!
```
Comment (cantino) [inline app/models/agents/google_calendar_publish_agent.rb]: Should this be in an `ensure` block?

Comment (albertsun) [inline app/models/agents/google_calendar_publish_agent.rb]: Ah yes, good idea

**Comment by albertsun** [comment]
Does this look mergeable now?

**Comment by dsander** [comment]
I think so, was waiting for second approval though.

@knu Could you have a third look at this?

File: app/controllers/application_controller.rb [hunk 6]
```diff
@@ -70,6 +71,12 @@ def outdated_docker_image_namespace_check
     @outdated_docker_image_namespace = ENV['OUTDATED_DOCKER_IMAGE_NAMESPACE'] == 'true'
   end
 
+  def outdated_google_auth_check
+    @outdated_google_cal_agents = current_user.agents.where("type like 'Agents::GoogleCalendarPublishAgent'").select do |agent|
```
Comment (knu) [inline app/controllers/application_controller.rb]: Or `of_type('Agents::GoogleCalendarPublishAgent')` to save a few more letters.

File: lib/google_calendar.rb [hunk 7]
```diff
@@ -1,33 +1,37 @@
...
   def initialize(config, logger)
     @config = config
 
     if @config['google']['key'].present?
-      @key = OpenSSL::PKCS12.new(@config['google']['key'], @config['google']['key_secret']).key
-    else
-      @key = Google::APIClient::PKCS12.load_key(@config['google']['key_file'], @config['google']['key_secret'])
+      # https://github.com/google/google-auth-library-ruby/issues/65
+      # https://github.com/google/google-api-ruby-client/issues/370
+      ENV['GOOGLE_PRIVATE_KEY'] = @config['google']['key']
```
Comment (knu) [inline lib/google_calendar.rb]: Isn't use of environment variables avoidable?  It'll potentially block our way to adopting a multi-threaded worker like Sidekiq in the future.

Maybe we should later introduce an exclusive lock to make sure the altered ENV values wouldn't affect (leak out to) other threads.

File: lib/google_calendar.rb [hunk 8]
```diff
@@ -56,14 +61,18 @@ def events_as(who, date)
...
+    ret = @calendar.list_events(
+        who
+      )
 
     @logger.debug ret.to_yaml
-    ret    
+    ret.to_h  
+  end
+
+  def cleanup!
```
Comment (knu) [inline lib/google_calendar.rb]: Why not provide a wrapper class method like this?
```ruby
  def self.open(*args, &block)
    instance = new(*args)
    block.call(instance)
  ensure
    instance.cleanup!
  end
```

File: lib/google_calendar.rb (hunk 7 above)
Comment (albertsun) [inline lib/google_calendar.rb]: Sadly it doesn't seem to be avoidable. Looking through the Google code there doesn't seem to be a way to pass these values in directly.

File: lib/google_calendar.rb (hunk 8 above)
Comment (albertsun) [inline lib/google_calendar.rb]: Hm. I don't think I understand that snippet of code. Does that change how the GoogleCalendar class gets called? 

Change this part?
 https://github.com/albertsun/huginn/blob/b4fb6e2e276cc8c5908f7d3f8e4257d24b298ca1/app/models/agents/google_calendar_publish_agent.rb#L111  to something like this?

```
calendar_event = GoogleCalendar.new(interpolate_options(options, event), Rails.logger).open |calendar|
   # stuff
   calendar.publish_as( ... )
end
```

Comment (knu) [inline lib/google_calendar.rb]: I meant you could rewrite it like this:
```ruby
incoming_events.each do |event|
  GoogleCalendar.open(interpolate_options(options, event), Rails.logger) do |calendar|
    cal_message = event.payload["message"]
    # ...
  end
end
```
This way the caller (GoogleCalendarPublishAgent#receive) doesn't need to bother with `cleanup!`, and the callee (GoogleCalendar.open) can easily manage thread safety because it takes a block in which a whole session is enclosed.

Comment (albertsun) [inline lib/google_calendar.rb]: Ohh cool, smart idea. Just added a commit to do that.

**Comment by albertsun** [comment]
Hey @knu @cantino @dsander how does this look now, mergeable?

**Comment by knu** [comment]
Looks good to me besides https://github.com/huginn/huginn/pull/2047#discussion_r128179012. 👍

**Comment by albertsun** [comment]
Right @knu, just did that, thanks!

**Comment by knu** [comment]
Perfect, thanks!

**Comment by albertsun** [comment]
Thanks @knu !
